datafusion/core/tests/sql/window.rs

```diff
@@ -1748,17 +1748,20 @@ async fn test_window_partition_by_order_by() -> Result<()> {
 
     let msg = format!("Creating logical plan for '{}'", sql);
     let dataframe = ctx.sql(sql).await.expect(&msg);
-    let physical_plan = dataframe.create_physical_plan().await.unwrap();
+    let physical_plan = dataframe.create_physical_plan().await?;
     let formatted = displayable(physical_plan.as_ref()).indent().to_string();
-    // Only 1 SortExec was added
     let expected = {
         vec![
-            "ProjectionExec: expr=[SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1, aggregate_test_100.c2] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@0 as SUM(aggregate_test_100.c4), COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@1 as COUNT(UInt8(1))]",
-            "  WindowAggExec: wdw=[SUM(aggregate_test_100.c4): Ok(Field { name: \"SUM(aggregate_test_100.c4)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }, COUNT(UInt8(1)): Ok(Field { name: \"COUNT(UInt8(1))\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
-            "    SortExec: [c1@0 ASC NULLS LAST,c2@1 ASC NULLS LAST]",
+            "ProjectionExec: expr=[SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1, aggregate_test_100.c2] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@1 as SUM(aggregate_test_100.c4), COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@0 as COUNT(UInt8(1))]",
+            "  WindowAggExec: wdw=[COUNT(UInt8(1)): Ok(Field { name: \"COUNT(UInt8(1))\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
+            "    SortExec: [c1@1 ASC NULLS LAST,c2@2 ASC NULLS LAST]",
             "      CoalesceBatchesExec: target_batch_size=4096",
-            "        RepartitionExec: partitioning=Hash([Column { name: \"c1\", index: 0 }], 2)",
-            "          RepartitionExec: partitioning=RoundRobinBatch(2)",
+            "        RepartitionExec: partitioning=Hash([Column { name: \"c1\", index: 1 }], 2)",
+            "          WindowAggExec: wdw=[SUM(aggregate_test_100.c4): Ok(Field { name: \"SUM(aggregate_test_100.c4)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
+            "            SortExec: [c1@0 ASC NULLS LAST,c2@1 ASC NULLS LAST]",
```

Comment:
In the previous version expressions below was going to the same `WindowAggExec`. 
```sql
     SUM(c4) OVER(PARTITION BY c1, c2 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING),
     COUNT(*) OVER(PARTITION BY c1 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) 
```
However, this behavior is contradictory to POSTGRE's behavior. When I run the query
```sql
SELECT 
               SUM(c4) OVER(PARTITION BY c1, c2 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING),
               COUNT(*) OVER(PARTITION BY c1 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) 
               FROM aggregate_test_100
```
POSTGRE produces physical plan below
```sql
WindowAgg (cost=34.94..53.34 rows=460 width=52) (actual time=0.112..0.243 rows=100 loops=1)
-> WindowAgg (cost=34.94..45.29 rows=460 width=46) (actual time=0.105..0.187 rows=100 loops=1)
-> Sort (cost=34.94..36.09 rows=460 width=38) (actual time=0.097..0.103 rows=100 loops=1)
Sort Key: c1, c2
Sort Method: quicksort Memory: 29kB
-> Seq Scan on aggregate_test_100 (cost=0.00..14.60 rows=460 width=38) (actual time=0.012..0.026 rows=100 loops=1)
```
The reason we cannot optimize away second `SortExec` is that. `RepartitionExec` doesn't maintain ordering. In the future, we plan to add an order preserving `RepartitionExec`, when this happens. Second `SortExec` will be unnecessary.

Reply:
ok -- I do think the original plan is better, but I would also be happy to merge this PR as is and improve it (potentially) as a follow on.

Reply:
I do think this is not a good optimization, it not only introduced another `SortExec`, but also another `RepartitionExec`. I will take a look and try to fix it in my refactoring PR.

Reply:
@mustafasrepo Could you please tell me the code place and the PR which lead to this behavior 
change ?

Reply:
Ok, I find out the place.  Could you please explain why we need to differ the sort key is from 
`PARTITION BY` or `ORDER BY `?

```rust
/// (expr, "is the SortExpr for window (either comes from PARTITION BY or ORDER BY columns)")
/// if bool is true SortExpr comes from `PARTITION BY` column, if false comes from `ORDER BY` column
type WindowSortKey = Vec<(Expr, bool)>;
```

Reply:
The reason is that ordering requirement by `ORDER BY` clause gives the required ordering for `WindowAggExec` to produce desired result. However, `PARTITION BY` ordering is only used to determine partition boundaries. Hence as long as we know that `PARTITION BY` columns are ordered (ordering direction is not important). We can calculate partition boundaries. This differentiation between `PARTITION BY` columns and `ORDER BY` columns enables us to produce better plans. With this differation we can produce more optimized plans for queries like following
```sql
SELECT
    c9, SUM(c9) OVER(ORDER BY c1, c9 DESC ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING) as sum1,
    SUM(c9) OVER(PARTITION BY c1 ORDER BY c9 DESC ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING) as sum2
    FROM aggregate_test_100
    LIMIT 5
```

Reply:
Sorry that I still do not understand.  For this SQL, could you please explain what a better plan it should be and what was the old plan ?
Yes, `PARTITION BY` columns are ordered and the ordering direction is not important, can we normalized that and make the sorting ordering to be consistent with the other window expressions and make them can be evaluated in one
`WindowAggExec`.

Reply:
I have changed the query to reflect use case better. 
`PARTITION BY` aware rule produces the plan below
```
"ProjectionExec: expr=[c3@1 as c3, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 DESC NULLS FIRST, aggregate_test_100.c9 DESC NULLS FIRST, aggregate_test_100.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@3 as sum1, SUM(aggregate_test_100.c9) PARTITION BY [aggregate_test_100.c3] ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@4 as sum2]",
"  GlobalLimitExec: skip=0, fetch=5",
"    BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c9): Ok(Field { name: \"SUM(aggregate_test_100.c9)\", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(UInt32(NULL)), end_bound: CurrentRow }]",
"      BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c9): Ok(Field { name: \"SUM(aggregate_test_100.c9)\", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int8(NULL)), end_bound: CurrentRow }]",
"        SortExec: [c3@1 DESC,c9@2 DESC,c2@0 ASC NULLS LAST]",
```
previously it was producing the plan below
```
"ProjectionExec: expr=[c3@1 as c3, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 DESC NULLS FIRST, aggregate_test_100.c9 DESC NULLS FIRST, aggregate_test_100.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@3 as sum1, SUM(aggregate_test_100.c9) PARTITION BY [aggregate_test_100.c3] ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@4 as sum2]",
"  GlobalLimitExec: skip=0, fetch=5",
"    BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c9): Ok(Field { name: \"SUM(aggregate_test_100.c9)\", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(UInt32(NULL)), end_bound: CurrentRow }]",
"      SortExec: [c3@1 ASC,c9@2 DESC]",
"        BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c9): Ok(Field { name: \"SUM(aggregate_test_100.c9)\", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int8(NULL)), end_bound: CurrentRow }]",
"          SortExec: [c3@1 DESC,c9@2 DESC,c2@0 ASC NULLS LAST]",
```
As you say, previously the query 
```
SELECT SUM(c4) OVER(PARTITION BY c1, c2 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING),
    COUNT(*) OVER(PARTITION BY c1 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) 
    FROM aggregate_test_100
```
was producing single `WindowAggExec`. However, with the change in this PR we no longer put window expressions `OVER(PARTITION BY c1, c2 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)` and `COUNT(*) OVER(PARTITION BY c1 ORDER BY c2 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)` to the same executor. Because of this change new rule produces less optimized plan for this specific change.
However, I think unless `PARTITION BY` columns and `ORDER BY` are exactly equal, different aggregations shouldn't end up in the same executor.

Reply:
I would like to contribute to the discussion by explaining the general motivation here: An important reason as to why `PARTITION BY` and `ORDER BY` are not treated the same is that we may want to utilize hashing instead of sorting for `PARTITION BY` keys in the (quite near) future. In applications where a user needs to run queries without breaking the pipeline (such as applications involving unbounded sources), hashing is the right choice; in other cases sorting is the right choice. So we really shouldn't overfit to treating them uniformly -- the right strategy depends on the situation.

Now, coming back to the specifics of this example: IMO, the root issue here is the intermediate `RepartitionExec`, which breaks the first sort and therefore induces the second sort. To solve this, we are working on sort-preserving repartitioning, which will solve problems like this -- the second sort will not come into the picture when we use the appropriate partitioning.

In the future, as we have more information available to us (e.g. are we using hashing or sorting for `PARTITION BY`s, does the raw input have a compatible ordering, what is our parallelism configuration, etc.), we can make the decision as to whether we want to have a single executor or multiple executors in cases like this. I would be definitely on board with that. However; the default choice (in the absence of these pieces of information) should be not to make any assumptions, so I agree with @mustafasrepo for the time being.